### PR TITLE
[Jobs] Add max_duration support for managed jobs

### LINKED ISSUE #10370: [Jobs] feature request: support `max_duration` setting
Hey, it sound like this (IMO quite fundamental) feature is currently missing. I would like to be able to configure something like: 

```yaml
job:
    max_duration: 10h
```

and Skypilot (server) should terminate the job on this timeout.

### PR DESCRIPTION
**Pull request by MitaliMSahoo**
Fixes skypilot-org/skypilot#10370

# Summary

Add support for a `max_duration` field in managed job YAMLs. When set, the SkyPilot jobs controller automatically terminates a managed job once it has been running longer than the configured limit (e.g. `max_duration`: 10h). This gives users a simple, declarative way to bound job runtime for cost control and reliability, a job that hangs or runs too long is killed automatically instead of running indefinitely.

## Example:

```
resources:
  cpus: 1
max_duration: 2h
run: |
  python train.py
```

# Description of Changes

* Task spec: `max_duration` is parsed from the YAML, validated at submission time (rejecting malformed values like `max_duration: banana` with a clear error rather than failing confusingly later), and round-trips through serialization like other task fields.
* Controller: the jobs controller's monitoring loop evaluates each task's elapsed time (from the persisted `start_at`) against its `max_duration` on every iteration, independent of the task's live status — so a task stuck in `RECOVERING` or hitting transient status-fetch errors is still subject to the deadline, not just one that's cleanly `RUNNING`. Applies correctly to every task in a pipeline or job group, not just the first. On timeout, the job is marked `FAILED` with an explicit reason ("Job exceeded `max_duration` of . The job was terminated by the controller.") and its cluster is torn down.
* CLI warning: sky launch / sky exec warn (rather than silently ignore) if max_duration is set, since the field is only enforced for managed jobs (sky jobs launch).
* Client/server version gating: `sky jobs launch`checks the connected API server's version before submitting a job with `max_duration` set, and fails fast with a clear error if the server predates support for the field, avoiding an opaque schema-validation rejection on older, not-yet-upgraded servers. Bumps `API_VERSION` per the project's backward-compatibility guidelines.
* Docs: added `max_duration` to the task YAML reference (docs/source/reference/yaml-spec.rst), including supported duration units and the managed-jobs-only scope.
* Backward compatible: jobs without `max_duration` set are unaffected.

# Tested

## What the tests verify:

* Unit tests cover the core timeout logic in isolation: a job with no `max_duration` never times out, a job within its limit does not time out, a job past its limit does time out, duration strings (`1m`, `90s`, `1d`) parse correctly, and the job start-time lookup returns the correct value (including edge cases like unknown/not-started tasks).
* The smoke test verifies the feature end-to-end on a real cluster: a managed job with `max_duration` set is launched, runs past its limit, and is terminated by the controller with the expected failure reason.

## Tests run:

* `bash format.sh` — yapf, isort, mypy, pylint all pass (mypy: no issues in 587 source files; pylint: 10.00/10)
* `pytest tests/unit_tests/test_jobs_max_duration.py` — 9 passed

  image
* Smoke test `test_managed_job_max_duration`, run locally against a Kubernetes cluster:

  image
* Manually verified end-to-end on a local Kubernetes (minikube) cluster: launched a managed job with `max_duration: 1m` and a 10-minute sleep, confirmed it was terminated by the controller around the 1-minute mark with the correct failure reason, and confirmed the underlying pod was actually cleaned up (not just marked failed).

  image

## Why this doesn't break existing jobs:

* `max_duration` is a new optional field (`Optional[str] = None`). Existing YAMLs without it are unaffected, and the controller's timeout check is guarded by `if task.max_duration is not None`, so jobs without it never hit the new logic.
* The timeout check itself returns `False` when `max_duration` is `None`, so existing managed jobs run exactly as before.
* The timeout marks the job `FAILED` through the same state-transition path used elsewhere, which only applies if the job hasn't already reached a terminal state, this avoids conflicting with cancellation or other terminal transitions happening concurrently.

---

Open in Devin Review

### CODE REVIEW

File: sky/jobs/controller.py [hunk 1]
```diff
@@ -158,6 +189,7 @@ def _build_task_specs(
     base_specs: Dict[str, Any] = {
         'max_restarts_on_errors': executor.max_restarts_on_errors,
         'recover_on_exit_codes': executor.recover_on_exit_codes,
+        'max_duration': executor.dag.tasks[executor.task_id].max_duration,
```
[the remark this answers is not in the document]
Comment (MitaliMSahoo) [reply sky/jobs/controller.py]: Fixed in the latest push: executor.dag.tasks[0] now used consistently, added a job-group regression test
Also addressed all Devin raised issues in the below list of commits.

**Comment by MitaliMSahoo** [comment]
Hi @DanielZhangQD / SkyPilot team - bumping this PR for review when you have a moment!

Status: Ready to merge
• Fixed all 4 Devin bot issues
• format.sh clean & 9/9 unit tests passing
• Verified via end-to-end local K8s smoke test

Happy to address any feedback or route elsewhere. Thanks!

**Comment by DanielZhangQD** [comment]
> #10370

Thanks for the support! @MitaliMSahoo 

cc @Michaelvll @cg505 for the UX change to the managed jobs.
